File: src/export/pageSpec.ts
/**
 * PageSpec — the page geometry that must travel to Rust, in points.
 *
 * Purpose: Windows and Linux ignore `@page { size }` entirely. Measured
 * (WI-PDF0.1): with no API geometry supplied at all, Windows emits US Letter
 * and Linux emits A4 — neither reads the CSS. So the size the user picked has
 * to reach the backend as data, not only as a stylesheet.
 *
 * Key decisions:
 *   - **Points, with orientation already applied.** Landscape is a WIDTH/HEIGHT
 *     SWAP, not a flag: Windows ignored
 *     `COREWEBVIEW2_PRINT_ORIENTATION_LANDSCAPE` while explicit width and
 *     height were set, and its MediaBox stayed portrait. Swapping is correct
 *     on both platforms.
 *   - **Margins travel too, and only Linux reads them.** The original decision
 *     was "no margins", from a probe that measured margins by PAGE COUNT. That
 *     signal is blind to the horizontal axis: left and right margins do not
 *     change how many pages a document needs. Measuring the ink box of a real
 *     export instead showed Linux printing edge to edge on ALL FOUR sides
 *     (0/4/4/0 pt where macOS and Windows both gave 72), because WebKitGTK
 *     takes its page box from the GtkPageSetup and ignores `@page { margin }`.
 *     macOS and Windows continue to honour the CSS and ignore these fields.
 *   - **One source, two derivations.** The CSS keeps using CSS keywords
 *     (`A4 landscape`) because `pdfHtmlTemplate.ts` records that an explicit
 *     length pair plus an orientation keyword is invalid and silently ignored
 *     by WebKit. Both this table and that one are keyed on the same
 *     `options.pageSize`, and `pageSpec.test.ts` fails if one gains an entry
 *     the other lacks.
 *
 * @coordinates-with pdfHtmlTemplate.ts — the CSS half of the same choice
 * @coordinates-with PdfExportDialog.tsx — builds this for the invoke payload
 * @module export/pageSpec
 */

/** Paper dimensions in PostScript points (1pt = 1/72in), portrait. */
export const PAGE_SIZE_PT: Record<string, { width: number; height: number }> = {
  a4: { width: 595.28, height: 841.89 }, // 210 × 297 mm
  letter: { width: 612, height: 792 }, // 8.5 × 11 in
  a3: { width: 841.89, height: 1190.55 }, // 297 × 420 mm
  legal: { width: 612, height: 1008 }, // 8.5 × 14 in
};

/** The geometry Rust receives. Field names are the wire contract. */
export interface PageSpec {
  widthPt: number;
  heightPt: number;
  marginTopPt: number;
  marginRightPt: number;
  marginBottomPt: number;
  marginLeftPt: number;
}

/** Millimetres to PostScript points. The dialog stores margins in mm. */
export function mmToPt(mm: number): number {
  return (mm * 72) / 25.4;
}

/**
 * Build the spec for a page size and orientation.
 *
 * Unknown sizes fall back to A4, matching `buildPageCSS`'s fallback — the two
 * must degrade the same way or a bad preset would produce a document whose CSS
 * and paper disagree.
 */
export function buildPageSpec(
  pageSize: string,
  orientation: string,
  margins?: { top: number; right: number; bottom: number; left: number },
): PageSpec {
  const paper = PAGE_SIZE_PT[pageSize] ?? PAGE_SIZE_PT.a4;
  const landscape = orientation === "landscape";
  // Margins are NOT swapped with the page. A left margin stays on the left
  // when the sheet turns; only the paper's width and height trade places.
  return {
    widthPt: landscape ? paper.height : paper.width,
    heightPt: landscape ? paper.width : paper.height,
    marginTopPt: mmToPt(margins?.top ?? 0),
    marginRightPt: mmToPt(margins?.right ?? 0),
    marginBottomPt: mmToPt(margins?.bottom ?? 0),
    marginLeftPt: mmToPt(margins?.left ?? 0),
  };
}

File: src-tauri/src/pdf_export/page_spec.rs
//! `PageSpec` — the page geometry the frontend sends, in points.
//!
//! Purpose: Windows and Linux ignore `@page { size }` entirely (ADR-PDF1), so
//! the size the user chose has to arrive as data. Measured with no API
//! geometry supplied at all, Windows emits US Letter and Linux emits A4 —
//! neither reads the stylesheet.
//!
//! Key decisions:
//!   - **Orientation is already applied.** Landscape is a width/height swap
//!     performed by the caller, not a flag: Windows ignored
//!     `COREWEBVIEW2_PRINT_ORIENTATION_LANDSCAPE` while explicit width and
//!     height were set (ADR-PDF1a).
//!   - **Margins: CSS on macOS and Windows, DATA on Linux.** The original
//!     "no margins" decision came from a probe that measured margins by PAGE
//!     COUNT — a signal blind to the horizontal axis, since left/right margins
//!     do not change how many pages a document needs. Rendering the real
//!     document and measuring the ink box showed Linux printing edge to edge
//!     on all four sides (0/4/4/0 pt against an expected 72). WebKitGTK's print
//!     path takes its page box from `GtkPageSetup`, which `linux.rs` was
//!     setting to zero on the strength of that probe. So margins now travel as
//!     data, and Linux is the only backend that reads them.
//!   - **Validated at the boundary.** A non-finite or absurd value reaching a
//!     native print API is a crash or a silently wrong document, and this is
//!     an IPC edge — zero trust applies.
//!
//! @coordinates-with src/export/pageSpec.ts — builds this
//! @module pdf_export/page_spec

use crate::command_error::{CommandError, ErrorCode};
use crate::localized_error;

/// Smallest sane page: below this nothing renders usefully.
const MIN_PT: f64 = 72.0; // 1 inch
/// Largest sane page. PDF's own ceiling is 14 400pt (200in); anything near it
/// is a bad unit conversion rather than a real page.
const MAX_PT: f64 = 14_400.0;

/// Page geometry in PostScript points, orientation already applied.
#[derive(Clone, Copy, Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PageSpec {
    pub width_pt: f64,
    pub height_pt: f64,
    /// Margins in points. `None` on a caller that predates them, which keeps
    /// the previous edge-to-edge behaviour rather than inventing a default.
    #[serde(default)]
    pub margin_top_pt: Option<f64>,
    #[serde(default)]
    pub margin_right_pt: Option<f64>,
    #[serde(default)]
    pub margin_bottom_pt: Option<f64>,
    #[serde(default)]
    pub margin_left_pt: Option<f64>,
}

impl PageSpec {
    /// Geometry with no margins — the shape every caller used before margins
    /// existed, and what the smoke fixtures still want.
    #[allow(dead_code)]
    pub const fn new(width_pt: f64, height_pt: f64) -> Self {
        Self {
            width_pt,
            height_pt,
            margin_top_pt: None,
            margin_right_pt: None,
            margin_bottom_pt: None,
            margin_left_pt: None,
        }
    }

    /// Reject geometry a native print API cannot honour.
    ///
    /// `is_finite` is the load-bearing check: `NaN` fails every comparison, so
    /// a range test alone would let it through and hand `NaN` to WebView2.
    pub fn validate(&self) -> Result<(), CommandError> {
        for (name, v) in [("width", self.width_pt), ("height", self.height_pt)] {
            if !v.is_finite() || !(MIN_PT..=MAX_PT).contains(&v) {
                return Err(localized_error!(
                    ErrorCode::InvalidInput,
                    "errors.pdf.invalidPageSize",
                    axis = name,
                    value = format!("{v}")
                ));
            }
        }
        // Margins are optional, but a supplied one must be sane: negative or
        // non-finite reaches GTK directly, and a pair wider than the page
        // leaves no content area at all.
        for (name, m, extent) in [
            ("top", self.margin_top_pt, self.height_pt),
            ("bottom", self.margin_bottom_pt, self.height_pt),
            ("left", self.margin_left_pt, self.width_pt),
            ("right", self.margin_right_pt, self.width_pt),
        ] {
            if let Some(v) = m {
                if !v.is_finite() || v < 0.0 || v >= extent {
                    return Err(localized_error!(
                        ErrorCode::InvalidInput,
                        "errors.pdf.invalidPageSize",
                        axis = name,
                        value = format!("{v}")
                    ));
                }
            }
        }
        if self.margin_left_pt.unwrap_or(0.0) + self.margin_right_pt.unwrap_or(0.0) >= self.width_pt
            || self.margin_top_pt.unwrap_or(0.0) + self.margin_bottom_pt.unwrap_or(0.0)
                >= self.height_pt
        {
            return Err(localized_error!(
                ErrorCode::InvalidInput,
                "errors.pdf.invalidPageSize",
                axis = "margins",
                value = "exceed page".to_string()
            ));
        }
        Ok(())
    }

    /// Width and height in inches — the unit WebView2's print settings take.
    ///
    /// Unused in the library build until WI-PDF2.1 calls it; the conversion is
    /// written and tested now because it is where a wrong page size would come
    /// from, and a units bug is far cheaper to catch here than through a
    /// remote render loop. `page_spec.test.rs` exercises it.
    #[allow(dead_code)]
    pub fn inches(&self) -> (f64, f64) {
        (self.width_pt / 72.0, self.height_pt / 72.0)
    }
}

#[cfg(test)]
#[path = "page_spec.test.rs"]
mod tests;
